scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SepRegion.scala

```diff
@@ -16,22 +19,107 @@ sealed trait SepRegionNonIndented extends SepRegion {
   override final def isIndented = false
 }
 
+sealed trait RegionDelim extends SepRegion
+sealed trait RegionNonDelimNonIndented extends SepRegionNonIndented
+
 case class RegionIndent(override val indent: Int, override val closeOnNonCase: Boolean)
-    extends SepRegionIndented
+    extends SepRegionIndented with RegionDelim
 
-case class RegionParen(canProduceLF: Boolean) extends SepRegionNonIndented
+case object RegionParen extends SepRegionNonIndented with RegionDelim
 
-case object RegionBracket extends SepRegionNonIndented
+case object RegionBracket extends SepRegionNonIndented with RegionDelim
 
 case class RegionBrace(override val indent: Int, override val indentOnArrow: Boolean)
-    extends SepRegionNonIndented with CanProduceLF
+    extends SepRegionNonIndented with CanProduceLF with RegionDelim
 
-case object RegionCaseExpr extends SepRegionNonIndented
+case object RegionCaseExpr extends RegionNonDelimNonIndented
 case class RegionCaseBody(override val indent: Int) extends SepRegionNonIndented with CanProduceLF
 
 case class RegionEnum(override val indent: Int) extends SepRegionNonIndented with CanProduceLF
 
 case class RegionIndentEnum(override val indent: Int) extends SepRegionIndented
 
 // NOTE: Special case for Enum region is needed because parsing of 'case' statement is done differently
-case object RegionEnumArtificialMark extends SepRegionNonIndented
+case object RegionEnumArtificialMark extends RegionNonDelimNonIndented
```

Comment:
This got a bit complicated now with all the branches. We could add some docs maybe with explanation of each region maybe?  Especially some of the less clear traits.


If you have time to do that of course.

Reply:
of course, will do.